cpp-package/example/inference/imagenet_inference.cpp

```diff
@@ -0,0 +1,597 @@
+  val_iter->SetParam("label_width", 1);
+  val_iter->SetParam("data_shape", data_shape);
+  val_iter->SetParam("preprocess_threads", data_nthreads_);
+  val_iter->SetParam("shuffle_chunk_seed", shuffle_chunk_seed_);
+
+  // set Batch parameters
+  val_iter->SetParam("batch_size", input_shape[0]);
+
+  // image record parameters
+  val_iter->SetParam("shuffle", true);
+  val_iter->SetParam("seed", seed_);
+
+  // set normalize parameters
+  val_iter->SetParam("mean_r", rgb_mean_[0]);
+  val_iter->SetParam("mean_g", rgb_mean_[1]);
+  val_iter->SetParam("mean_b", rgb_mean_[2]);
+  val_iter->SetParam("std_r", rgb_std_[0]);
+  val_iter->SetParam("std_g", rgb_std_[1]);
+  val_iter->SetParam("std_b", rgb_std_[2]);
+
+  // set prefetcher parameters
+  if (use_gpu_) {
+    val_iter->SetParam("ctx", "gpu");
+  } else {
+    val_iter->SetParam("ctx", "cpu");
+  }
+  val_iter->SetParam("dtype", data_layer_type_);
+
+  val_iter->CreateDataIter();
+  return true;
+}
+
+/*
+ * The following function loads the model from json file.
+ */
+void Predictor::LoadModel(const std::string& model_json_file) {
+  if (!FileExists(model_json_file)) {
+    LG << "Model file " << model_json_file << " does not exist";
+    throw std::runtime_error("Model file does not exist");
+  }
+  LG << "Loading the model from " << model_json_file << std::endl;
+  net = Symbol::Load(model_json_file);
+}
+
+
+/*
+ * The following function loads the model parameters.
+ */
+void Predictor::LoadParameters(const std::string& model_parameters_file) {
+  if (!FileExists(model_parameters_file)) {
+    LG << "Parameter file " << model_parameters_file << " does not exist";
+    throw std::runtime_error("Model parameters does not exist");
+  }
+  LG << "Loading the model parameters from " << model_parameters_file << std::endl;
+  std::map<std::string, NDArray> parameters;
+  NDArray::Load(model_parameters_file, 0, &parameters);
+  for (const auto &k : parameters) {
+    if (k.first.substr(0, 4) == "aux:") {
+      auto name = k.first.substr(4, k.first.size() - 4);
+      aux_map[name] = k.second.Copy(global_ctx);
+    }
+    if (k.first.substr(0, 4) == "arg:") {
+      auto name = k.first.substr(4, k.first.size() - 4);
+      args_map[name] = k.second.Copy(global_ctx);
+    }
+  }
+  /*WaitAll is need when we copy data between GPU and the main memory*/
+  NDArray::WaitAll();
+}
+
+/*
+ * The following function randomly initializes the parameters when benchmark_ is true.
+ */
+void Predictor::InitParameters() {
+  std::vector<mx_uint> data_shape;
+  for (index_t i=0; i < input_shape.ndim(); i++) {
+    data_shape.push_back(input_shape[i]);
+  }
+
+  std::map<std::string, std::vector<mx_uint> > arg_shapes;
+  std::vector<std::vector<mx_uint> > aux_shapes, in_shapes, out_shapes;
+  arg_shapes["data"] = data_shape;
+  net.InferShape(arg_shapes, &in_shapes, &aux_shapes, &out_shapes);
+
+  // initializer to call
+  Xavier xavier(Xavier::uniform, Xavier::avg, 2.0f);
+  index_t i = 0;
```

Comment:
Why declare `i`?

Reply:
This is for indexing infered shape of each arguments (see Line334), which is needed for `Xavier` to initialize. So does `i = 0` in Line339.

Reply:
Quite strange usage. If so, please declare i in for loop header.

Reply:
Done